### ISSUE TITLE
Why is postfix  not in docker container ?

### ISSUE THREAD

**Issue by Neilpang** [simple-login/app#201]
You guys are doing a great job.  I wish I found this project earlier.

Just a small question,  why not run postfix in a container ?

I just want to dockerise all the services,  I don't want to install any service on my docker host.

**Comment by nguyenkims** [comment]
@Neilpang we want to dockerize Postfix as well but haven't found a reliable-enough Docker image for Postfix. I just created this ticket in our roadmap, feel free to vote on it to make it happen sooner :) https://trello.com/c/G5oBKex2/78-dockerize-postfix

**Comment by Neilpang** [comment]
Thanks for your reply.
what do you mean by "reliable-enough"?
Isn't it just a "apt-get install postfix"?
Or do you encounter some issues ?

Thanks.

**Comment by nguyenkims** [comment]
@Neilpang Docker container works a bit differently to a normal Linux server, for example it requires having a single process (among other things). As Postfix is composed of different processes, dockerizing Postfix is not obvious. Some people have tried some workarounds to create a Docker image for Postfix but these images are either not battle-tested and/or not maintained actively.

**Comment by arugifa** [comment]
@Neilpang If it can help, I created a small Postfix image for Simple Login : https://hub.docker.com/r/arugifa/simplelogin-postfix

I use it for a couple of months now, and don't have any major problem with it (i.e., I receive all emails forwarded by Simple Login).

The SSL connection is not working properly though on my setup, and I don't plan to fix it before next winter :snowman:

**Comment by nguyenkims** [comment]
@arugifa I just checked out your Docker image and it looks great! Do you think we can add that to the self-hosting instruction as it would vastly simplify the Postfix self-hosting section? 

Now that everything can be run as Docker container, we can also have a single `docker-compose.yml` file that runs all the necessary components, one just needs to setup a DNS to have their own SimpleLogin server.

**Comment by Neilpang** [comment]
That's great, man.

**Comment by arugifa** [comment]
> @arugifa I just checked out your Docker image and it looks great! Do you think we can add that to the self-hosting instruction as it would vastly simplify the Postfix self-hosting section?
> 
> Now that everything can be run as Docker container, we can also have a single `docker-compose.yml` file that runs all the necessary components, one just needs to setup a DNS to have their own SimpleLogin server.

Sure, would be really great :) It would make people life easier for sure.

It would just be necessary to add a warning regarding TLS connection to Postfix server. On my personal setup, it doesn't work: Let's Encrypt cannot access to the Certbot server when trying to resolve the HTTP challenge. I don't know yet if it's because of my Docker image, or the configuration of my reverse proxy (Traefik). I plan to solve this problem, but not before next fall/winter :+1:

**Comment by nguyenkims** [comment]
@arugifa thanks for the info!

**Comment by Metabaron1** [comment]
Hi
I also  checked Arugifa simplelogin-postfix container. It's working fine. But I'm wondering why certbot is linked to postfix container. Certbot  is used in standalone mode and http access from letsencrypt server is required to validate the domain. Then wouldn't it make more sense Certbot to be linked to Nginx container instead of Postfix? And just share letsencrypt container between Postfix and Nginx containers. Another reason is that Certbot can auto setup Nginx security parameters...

**Comment by nguyenkims** [comment]
@Metabaron1 by default Postfix uses a self-signed certificate which works but using a certificate issued by a known organization (Letsencrypt in case of Certbot) is always better :).

**Comment by Metabaron1** [comment]
Postix would continue to use Letsencrypt certificate with a shared letsencrypt volume + a "depends_on" nginx ;)
 I'm just suggesting moving Certbot certificate update  from Postix container to Nginx continer which already have an opened http port.

**Comment by arugifa** [comment]
Hi @Metabaron1 ,

Thanks for your feedback :+1: 

It is a valid point that you raised, but I believe there is a problem with your approach: you are assuming that some Nginx server should be responsible to generate a Let's Encrypt certificate for Simple Login web app. However, in my case for example, I am using Simple Login in addition of many other services. Each of them running behind a reverse load balancer named [Traefik](https://github.com/containous/traefik/).

This load balancer is responsible to generate Let's Encrypt certificates, so I don't need any Nginx server to do the same job. Also, what if a user wants to have two different domains, and thus two different certificates: one for the Simple Login web app, and one for Postfix. How would you handle such case?

Right now, running Certbot inside Postfix container seems more flexible to me. But I may be wrong.

A viable option could be that I make Certbot optional in my Postfix image. So people could adapt it to their use case.

**Comment by Metabaron1** [comment]
Hi @arugifa
Yes, I noticed your Traefik, and then understood that that's why you didn't need Nginx that is only used as proxy within SimpleLogin. That's why I commented here and not in your project, as it's in the roadmap to make a single docker-compose for all SimpleLogin components.

But your are rigth, I was wrong assuming the same domain for mail and http access. In this case, I suppose it's indeed better to have 2 Certbots : 1 within Nginx container, 1 within Postfix container with the option to disable it.

Or another solution, for those not using Traefik, Nginx with multiple vhost for each domain mail ;)
